ocaml/xapi/xapi_network.ml

```diff
@@ -329,3 +329,62 @@ let with_networks_attached_for_vm ~__context ?host ~vm f =
         error "Caught %s while detaching networks" (string_of_exn e)
     end;
     raise e
+
+let assert_can_add_purpose ~__context ~network ~current porpoise =
+  let sop (*string-of-porpoise*) = Record_util.network_purpose_to_string in
+  let reject str =
+    raise (Api_errors.Server_error (Api_errors.network_incompatible_purposes, [ (sop porpoise); str ]))
+  in
+  if (List.mem `himn current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `himn);
+    reject "any"
+  );
+  if (List.mem `unmanaged current) then (
+    info "Existing network purpose %s is not compatible with any other purpose." (sop `unmanaged);
+    reject "any"
+  );
+
+  let assert_no_net_has_bad_porpoise bads =
+    (* Sadly we can't use Db.Network.get_refs_where because the expression
+     * type doesn't allow searching for a value inside a list. *)
```

Comment:
I don't think that's sad, because this code is more readable then get_..._where and also typesafe 😄

Reply:
This way takes more computation and memory, but you make a fair point.